### DISCUSSION TITLE
Midi on 1st core - webserver on 2nd core?

### DISCUSSION THREAD

**Discussion by fapplin** [earlephilhower/arduino-pico#727] [Q&A]
Hi,

First, thanks for all the great work on this tremendous endeavor!!! 

Second, my question - can I run midi on 1 core and a web server on the other core? My goal is pretty simple. The web server will be gathering a few values (and not very often) and the midi will use the web server values. 

Normally, you'd do thread with locking. This is why why was looking for AP Mode via a phone. I basically want to plug in a usb attached to the Pico W (it's recognized as a midi device) and the midi part is working. I'd like to be able to connect via my phone (no wifi router) and set the few values via a small web page. 

Thanks,

Frank

**Comment by fapplin** [comment]
OK. Got AP web server working on Core 1. Note: you have to turn off your 4G/5G service on your phone to use the AP mode. And, got MIDI working on Core 0.

**Reply by earlephilhower** [reply]
Awesome!  Is it something you think should be part of the examples here?  You can add a PR if you'd like.

**Reply by fapplin** [reply]
Well, this is for a money maker, but I'll see if I can come up with a good example for you of connecting to AP mode and turning on and off a couple LEDs (to keep it simple).

**Reply by fapplin** [reply]
OK. Here's a working example with a WiFi connection using AP Mode (I'm using an old android phone with no SIM card to access browser on phone). The WiFi connection should show up as PICOW and the password is "123456789".  Open a browser on the phone and enter the URL address is 192.168.42.1. and hit <Enter>. The web page has two buttons "LED On" and LED Off" - it will also update a label on the web page with the LED status. The LED is on GPIO 7. The WiFi part is on Core 1 and LED part is on Core 0.

```
//web server with WiFi in AP mode

#include <Arduino.h>
#include <WiFi.h>

#define LED1 7

#ifndef STASSID
#define STASSID "PICOW"
#define STAPSK "123456789"
#endif

const char* ssid = STASSID;
const char* password = STAPSK;
int port = 80;
char buf[4096];

WiFiServer server(port);

void updateWebpage(uint32_t led1)
{
  char message[128];
  
  strcpy(message, "");
  sprintf(message, "LED1: %d", led1);
  
  strcpy(buf, "<!DOCTYPE html>");
  strcat(buf, "<html>");
  strcat(buf, "<title>AP Mode Server Test</title>");
  strcat(buf, "<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">");
  strcat(buf, "<body>");
  strcat(buf, "<form id=\"form_kisi\" method=\"get\" action=\"/\">");
  strcat(buf, "<table >");
  strcat(buf, "<tr>");
  strcat(buf, "<td>");
  strcat(buf, "<div >");
  strcat(buf, "<button name=\"led_on\" type=\"submit\" >LED On</button>");
  strcat(buf, "</div>");
  strcat(buf, "</td>");
  strcat(buf, "<td>");
  strcat(buf, "<div >");
  strcat(buf, "<button name = \"led_off\" type=\"submit\"  >LED Off</button>");
  strcat(buf, "</div>");
  strcat(buf, "</td>");
  strcat(buf, "</tr> ");
  strcat(buf, "<tr><td><label>");
  strcat(buf, message);
  strcat(buf, "</label></td></tr>");
  strcat(buf, "</table>");
  strcat(buf, "</form>");  
  strcat(buf, "</body>");
  strcat(buf, "</html>");

}

void setup() {
  pinMode(LED1, OUTPUT);

}

void loop() {

  static uint32_t led1;

  if (rp2040.fifo.available() > 0) {
    led1 = rp2040.fifo.pop();
    Serial.print("led1: ");
    Serial.println(led1);
  } //end if

  if (led1 == 1)
  {
    digitalWrite(LED1, HIGH);
  }
  else
  {
    digitalWrite(LED1, LOW);
  }

}

void setup1() 
{
  WiFi.mode(WIFI_AP);
  WiFi.setHostname("PicoW2");
  WiFi.begin(ssid, password);
  delay(5000);

  server.begin();
}

void loop1() 
{
  static uint32_t led1;

  // Check if a client has connected
  WiFiClient client = server.available();
  String ip_addr = WiFi.localIP().toString().c_str();
  
  String req = client.readStringUntil('\n');
  if (client.available()) {  
    if (req.indexOf("led_on") > 0 ) 
      led1 = 1;
    else if (req.indexOf("led_off") > 0 )
      led1 = 0;
      
    rp2040.fifo.push(led1);
      
    updateWebpage(led1);
  } //end if 


  client.flush();
  client.print(buf);
  strcpy(buf, "");

  delay(10);

}
```
